### ISSUE TITLE
Authorization header not being set when provided as [HeaderCollection]

### ISSUE THREAD

**Issue by DrLeh** [reactiveui/refit#1374]
My issue stated here: https://stackoverflow.com/questions/72172657/c-sharp-refit-client-is-not-sending-authorization-header

I'm using a C# client to have my services talk to one another via http.

I'm trying to send a Bearer token through the Authorization header, but according to the error message, it's not setting the AZ header in the request (see bottom). I've tried setting it through providing all headers, and using the [Authorize] attribute, and all the other methods as described in the readme.

Here is my Refit client api call definition:

```c#
[Post(PresentationsBasePath + "/{presentationId}/cart")]
Task AddItemToCartAsync(long presentationId, ShoppingCartItemView item, [HeaderCollection] IDictionary<string, string> headers);

private Dictionary<string, string> GetTokenHeader(long presentationId, string token) => new()
{
    ["pres_id"] = presentationId.ToString(),
    [HeaderNames.Authorization] = $"Bearer {token}",
};

var header = GetTokenHeader(presentationId, token);
await _api.AddItemToCartAsync(presentationId, item, header);

```

However, I'm getting a 401, and looking at the Refit.ApiException that's thrown, the RequestMessage.Headers does not contain the Authorization header.  When I debug the receiving server, i'm seeing that there is indeed no Authorization header set on the incoming request.

Here's how I'm registering my refit api `IPresentationsApi`. I'm not doing anything relating to auth in the DI configuration

```c#
var refitSettings = GetRefitSettings();

void Configure<T>() where T : class => services
    .AddRefitClient<T>()
    .ConfigureHttpClient(ConfigureHttpClient);

Configure<IMarsPresentationApi>();
//other apis configured below

    
    private static void ConfigureHttpClient(IServiceProvider sp, HttpClient client)
    {
        var config = sp.GetRequiredService<IMarsConfiguration>();
        if (config.BaseUrl == null)
            throw new InvalidOperationException("Mars:BaseUrl must be configured");
        client.BaseAddress = new Uri(config.BaseUrl);
    }
```
Error shown here- you can see I get 401 because the AZ header is not set in the request as expected, even though the `pres_id` value is:

![image](https://user-images.githubusercontent.com/1790357/172906127-59ed5cbd-ecd2-491f-8936-2502a742cd25.png)

And on the server-side I can also verify that there's no AZ header set
![image](https://user-images.githubusercontent.com/1790357/172906523-3c094694-55a1-4498-8b82-83842353366a.png)

What am I doing wrong? How do I get it to send the AZ header?

**Environment**
- Refit Version:   6.3.2

**Comment by james-s-tayler** [comment]
From your code sample it doesn't look like you're doing anything wrong, so it's surprising you say it's not working. I had a go at doing a reproduction of this and wrote some tests including one for this specific scenario but all the tests pass and when I inspect it via the debugger it works perfectly. 

https://github.com/reactiveui/refit/pull/1385

**Comment by supermitsuba** [comment]
Im having the same issue, except I have narrowed down mine to adding it to refit Settings

`
using Microsoft.Net.Http.Headers;
using Refit;

var header = "<insert>";
var refitSettings = new RefitSettings()
{
    AuthorizationHeaderValueGetter = () => Task.FromResult(header)
};

var client = RestService.For<IGraphQLApi>("https://thegraphqlendpoint", refitSettings);
var result = await client.SendQuery("Doesnt matter");
Console.WriteLine(result.StatusCode);
Console.ReadKey();

public interface IGraphQLApi
{
    // for testing locally
    [Post("/GraphQL")]
    public Task<HttpResponseMessage> SendQuery([Body] string query);
}
`

I have used this same code, slightly changed in the past, and it has worked, but this time it is not working.  However, DrLeh way works for me.  Could be that Task.FromResult, but again it has worked in the past.  This is a .net core console application.

![refit](https://user-images.githubusercontent.com/2320336/181636044-5a953a81-3829-4d6b-b2c6-d12d22cb8d4c.png)

**Comment by b1ubb0r** [comment]
I had the same problem and searching for a solution I stumbled over several posts, saying that Authorization header gets stripped, when a redirect happens. 
https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.allowautoredirect?view=net-6.0#remarks
I changed my code to:
`        
services.AddHeaderPropagation(options => options.Headers.Add("Authorization"));
services.AddRefitClient<IIdentityApi>(settings)
                .ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler() { AllowAutoRedirect = false })
                .ConfigureHttpClient(client =>
                {
                    client.BaseAddress = new Uri(url);
                })
                .AddHeaderPropagation() // Enable on this specific client
                ;
`
With that I got an exception thrown, that a redirect happened.
Might be worth checking if it is the same for you.

**Comment by osnipezzini** [comment]
Same problem happening here

**Comment by simonneedham** [comment]
I hit this same issue and discovered that I was receiving redirects from my target API thanks to @b1ubb0r 's comment above ☝🏻. 

I wanted to add that in my case I was being returned a 301-Moved Permanently because I had configured my endpoint to be `/{presentationId}/cart` and this particular API was being a bit fussy and redirecting me to `/{presentationId}/cart/` 🙄

**Comment by meysambagheri96** [comment]
> I had the same problem and searching for a solution I stumbled over several posts, saying that Authorization header gets stripped, when a redirect happens. https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.allowautoredirect?view=net-6.0#remarks I changed my code to:
> 
> ```
> services.AddHeaderPropagation(options => options.Headers.Add("Authorization"));
> services.AddRefitClient<IIdentityApi>(settings)
>                 .ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler() { AllowAutoRedirect = false })
>                 .ConfigureHttpClient(client =>
>                 {
>                     client.BaseAddress = new Uri(url);
>                 })
>                 .AddHeaderPropagation() // Enable on this specific client
>                 ;
> ```
> 
> With that I got an exception thrown, that a redirect happened. Might be worth checking if it is the same for you.
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/64d54100868cbb4cbc950764ebd3bcd1a8075caf99f4b9184662880b6c232920/68747470733a2f2f646f63732e6d6963726f736f66742e636f6d2f656e2d75732f6d656469612f6c6f676f732f6c6f676f2d6d732d736f6369616c2e706e67">**[HttpClientHandler.AllowAutoRedirect Property (System.Net.Http)](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.allowautoredirect)**Gets or sets a value that indicates whether the handler should follow redirection responses.
> 
> Edit: fixed code formatting

It works for me. thanks!

**Comment by pankajverma82** [comment]
I am having same problem. Authorization header is getting stripped when redirect call happens. I don't think, I can use AllowAutoRedirect = false.

**Comment by TimothyMakkison** [comment]
Hey @pankajverma82 does @b1ubb0r comment help? Otherwise could you share your code.

**Comment by pankajverma82** [comment]
Not really, @TimothyMakkison As in my case server side api does redirect so I can't set AutoRedirect to false. I need to find way to pass the Auth header in 2'nd call (while its being redirected). **The issue is happening only with APIs where redirect happens, other api calls are working fine where there is no redirect from server side.** 

 public class AuthHeaderHandler : DelegatingHandler
 {
     private readonly IAuthTokenStore _authTokenStore;

     public AuthHeaderHandler(IAuthTokenStore authTokenStore)
     {
         _authTokenStore = authTokenStore;
     }
     protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
     {
         var token = await _authTokenStore.GetToken();
         request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token.Access_Token);
         return await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
     }
 }


**In Program.cs**

 webApplicationBuilder.Services.AddRefitClient<IViperRefitClient>(xmlRefitSettings)
  .ConfigureHttpClient(c => c.BaseAddress = new Uri(viperClient.BaseAddress))
 .**AddHttpMessageHandler**<AuthHeaderHandler>();

**Comment by glennawatson** [comment]
Closing as resolved/by-design. [HeaderCollection] with an IDictionary<string,string> containing "Authorization" does set the header correctly - this is covered by tests in AuthenticatedClientHandlerTests (HeaderCollection + Authorize/token scenarios) and verified against current main.

The 401s reported here are the well-known HttpClient behavior, confirmed by multiple commenters in this thread: the Authorization header is stripped on redirect (HttpClientHandler.AllowAutoRedirect). When the server 301/302-redirects (often just a missing/extra trailing slash), the auth header is dropped on the followed request. Fixes:
- correct the URL so no redirect happens, or
- set AllowAutoRedirect = false and target the final URL, or
- re-apply the header on redirect via a DelegatingHandler / header propagation.

If you have a case where the header is dropped with NO redirect involved, please open a fresh issue with a minimal repro repo, exact Refit version, and the captured raw request/response.
